### ISSUE TITLE
VS Code Support

### ISSUE THREAD

**Issue by MattMorgis** [mde/ejs#293]
This is more to start a community discussion, but support for [VS Code](https://code.visualstudio.com) is lacking. There is [one plugin](https://marketplace.visualstudio.com/items?itemName=QassimFarid.ejs-language-support) out there, but it's just [a fork of TextMate support](https://github.com/gregory-m/ejs-tmbundle) for EJS. 

It looks like VS Code needs a `.tmLanguage` file to offer support for a language. This a grammar format TextMate uses to identify keywords, comments, control flow, etc. when rendering an `.ejs` file. The [documentation](https://manual.macromates.com/en/language_grammars) seemed vague to me. This is why the above and only EJS extension forked the TextMate `.tmLanguage` file.

I have had issues with EJS support for a while. I moved from Sublime Text to Atom because of the EJS support [breaking in Sublime](https://github.com/samholmes/EJS.tmLanguage/issues/14#issuecomment-194145793). The best support I've seen for EJS is in WebStorm via a [Java plugin](https://plugins.jetbrains.com/plugin/7296-ejs) from JetBrains.

I've found a list of various `.tmLanguage` files for EJS and tried them all in VS Code. They all have various features that _kind of_ work. **I am proposing we build one officially supported `.tmLanguage` file that can be shared by Sublime, VS Code, TextMate, and any other editor that supports this format** (I _think_ Atom and Webstorm also do).

I'm not sure if this is something best suited for the official EJS language repo or organization or done as a community project. I'm willing to help out and even do a bulk of the work, it's when I got into the guts of it that things got tricky and I needed some help and brainstorming. Hence me reaching out to the community to see if there is already something in works, interest in getting this off the ground, etc. 

List of `EJS.tmLanguage` files:
* [@samholmes](https://github.com/samholmes/EJS.tmLanguage/blob/master/EJS.tmLanguage) - most elaborate I've found, used by and currently broken in Sublime
* [tmBundles](https://github.com/malomalo/tmbundles/blob/master/JavaScript.tmbundle/Syntaxes/HTML%20(Javascript%20-%20EJS).tmLanguage) - repo of text mate language bundles
* [@gregory-m](https://github.com/gregory-m/ejs-tmbundle/blob/master/Syntaxes/JavaScript%20Template.tmLanguage) - current VS Code extension built off of this and doesn't provide much support

Please let me know thoughts, next steps, or anything I'm missing or overlooking. 

In my opinion, these would be 'must haves':
* File `include`'s
* Comments
* JS Control Flow (if/else/loops) between closing `%>` and opening `<%`
* HTML in-line values. Ex. `<img src="<%= ejs %>">`

Current Screenshot of EJS in VS Code:
![vscode-ejs](https://user-images.githubusercontent.com/4664374/29005557-ad5a7942-7aab-11e7-95e5-bccf86f2a046.png)

Current Screenshot of EJS in WebStorm:
![webstorm-ejs](https://user-images.githubusercontent.com/4664374/29005561-b58d0418-7aab-11e7-96eb-3919e0902351.png)

**Comment by RyanZim** [comment]
This should be community-supported IMO. We could consider linking to a community project from the README perhaps.

**Comment by mde** [comment]
Community-supported would be ideal -- it seems like a non-trivial project unto itself. We're definitely happy to help and provide support however possible.

**Comment by ezekg** [comment]
Can we just fork [the `.erb` package](https://marketplace.visualstudio.com/items?itemName=CraigMaslowski.erb) and replace the Ruby-specific matching with JS? The `.erb` language seems to look great for `.ejs` files as-is, honestly. I have the following in my VS Code preferences:

```json
// Place your settings in this file to overwrite the default settings
{
    "files.associations": {
        "*.ejs": "erb"
    }
}
```

**Comment by mde** [comment]
Interesting idea! Would you like to take a crack at an adapted-for-JS fork?

**Comment by ezekg** [comment]
I gave it a try at [vscode-ejs](https://github.com/ezekg/vscode-ejs), but I can't seem to get JavaScript's control flow and parenthesis matching to stay within the `<%` and `%>` patterns—it seems that [you can't limit the scope of embedded language](https://github.com/Microsoft/vscode/issues/20488) so I've given up for now.

**Comment by mde** [comment]
Of course, it's two (or is it three?) interleaved grammars. Seems like a really interesting problem to solve. :)

**Comment by MattMorgis** [comment]
@ezekg That is the exact point that I created this issue! I had a similar thought to yours where I wanted to dig up Handlebars support and see if it handles this in any way that we could fork or re-implement, but it's still on my to-do list unfortunately. 

I wonder if there are any template engines who have done something similar to what you've mentioned already? I honestly don't even know what's out there these days - a quick Google brought up [this list](https://colorlib.com/wp/top-templating-engines-for-javascript/) which I can try to look into.

I honestly feel like we're _this_ close.

**Comment by markovicdenis** [comment]
Right now this extension (EJS language definition) is utterly useless. Wrong syntax highlighting when you use IF statements and whats worse, it doesnt mark closing tags in the html.

**Comment by Betanu701** [comment]
All,

I have created a new EJS Language definition. If you want to test it out and let me know if you have any issues. https://github.com/Betanu701/ejs-grammar

For now, if you open VSCode installation directory, find the other languages and create a new folder called "ejs" and place the code from my repo in there. I will be working on getting it integrated into VSCode. But you should now have access to select "Javascript EJS" from the languages.

I tried to get as much as I could in it. Let me know if I should add anything.

**Comment by Betanu701** [comment]
I have made this an actual extension in VSCode, Search ejs, the name is "Javascript EJS" Let me know if there are issues.

**Comment by MattMorgis** [comment]
@Betanu701 👏 👍 😄 

I'll install this and check it out a bit later! Very excited!

**Comment by ezekg** [comment]
You're awesome, @Betanu701. Looks great for me. 👏

**Comment by mde** [comment]
This is awesome. :) Would be great to get this added to the README and Web site.

**Comment by bombillazo** [comment]
This is desperately needed for `ejs`...

**Comment by RedCMD** [comment]
here's my VSCode extension to help with authoring TextMate grammars https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter

**Comment by mde** [comment]
This issue has been closed as outdated. 

If you believe this issue is still relevant, please:
1. Open a new issue with current context
2. Reference this issue number (#293)
3. Include details about your use case and EJS version

Thank you for your contribution!
